File: tools/prepare_refs.rs
use std::{env, process::ExitCode};

use rinha_backend_2026::ReferenceSet;

fn main() -> ExitCode {
    match run() {
        Ok(()) => ExitCode::SUCCESS,
        Err(error) => {
            eprintln!("{error}");
            ExitCode::FAILURE
        }
    }
}

fn run() -> Result<(), String> {
    let mut args = env::args().skip(1);

    let input_path = args.next().ok_or_else(usage)?;
    let references_bin = args.next().ok_or_else(usage)?;
    let labels_bin = args.next().ok_or_else(usage)?;

    if args.next().is_some() {
        return Err(usage());
    }

    let refs = ReferenceSet::load_gzip_json(&input_path)?;
    refs.write_binary(&references_bin, &labels_bin)?;

    eprintln!(
        "prepared {} referências em {} e {}",
        refs.len(),
        references_bin,
        labels_bin
    );

    Ok(())
}

fn usage() -> String {
    "uso: prepare_refs <references.json.gz> <references.bin> <labels.bin>".to_string()
}

File: c/src/prepare_refs.c
#include "rinha.h"

#include <stdio.h>

int main(int argc, char **argv) {
    if (argc != 4) {
        fprintf(stderr, "uso: prepare_refs_c <references.json.gz> <references.bin> <labels.bin>\n");
        return 1;
    }

    char error[256] = {0};
    ReferenceSet refs;
    if (!refs_load_gzip_json(argv[1], &refs, error, sizeof(error))) {
        fprintf(stderr, "%s\n", error);
        return 1;
    }

    if (!refs_write_binary(&refs, argv[2], argv[3], error, sizeof(error))) {
        fprintf(stderr, "%s\n", error);
        refs_free(&refs);
        return 1;
    }

    fprintf(stderr, "prepared %zu referências em %s e %s\n", refs.rows, argv[2], argv[3]);
    refs_free(&refs);
    return 0;
}
